onnxruntime/core/graph/graph_utils.h

```diff
@@ -186,5 +186,60 @@ void FinalizeNodeFusion(Graph& graph, Node& first_node, Node& second_node);
 */
 void FinalizeNodeFusion(Graph& graph, const std::vector<std::reference_wrapper<Node>>& nodes, Node& replacement_node);
 
+
+/*
+Find the first input edge of a node that the source node's operator type, version, and domain match the given values.
+@returns nullptr when not found.
+*/
+const Node::EdgeEnd*
+FindFirstInputEdge(
+    const Node& node,
+    const std::string& op_type,
+    const std::initializer_list<ONNX_NAMESPACE::OperatorSetVersion>& versions,
+    const std::string& domain);
+
+/*
+Find the input edge of a node for a specified input index.
+@returns nullptr when not found.
+*/
+const Node::EdgeEnd* GetInputEdge(const Node& node, int arg_index);
+
+/*
+Find the source node of an input edge for a specified input index.
+@returns nullptr when not found.
+*/
+const Node* GetInputNode(const Node& node, int arg_index);
+
+struct MatchEdgeInfo {
+  int dst_arg_index;
+  std::string op_type;
+  std::initializer_list<ONNX_NAMESPACE::OperatorSetVersion> versions;
+  std::string domain;
+};
+
+/*
+Find a path that matches the specified edge information.
+@param match_edges has information of a sequence of edges in the path.
+For example, the first is input edge for current node, and the second one is for the parent node.
+@param result stores edges that are found.
+@returns true when all edges are found.
+@remarks one limitation of this function is that it will not look at all possible paths when dst_arg_index=-1 is used.
+This is by design to reduce the complexity (Otherwise, recursive or stack is needed to enumerate all possible paths).
+Here is an example graph with two paths:
+      Add      Sub
+       |        |
+     Shape    Shape
+        \      /
+         Concat (current node)
+When you use [{dst_arg_index=1, op_type="Shape"}, {dst_arg_index=0, op_type="Sub"}], it will find the correct path.
+When you use [{dst_arg_index=-1, op_type="Shape"}, {dst_arg_index=0, op_type="Sub"}], it will report that path is not found.
+During mathing "Shape" for dst_arg_index=-1, it will get the edge of the first Shape node (with smaller node index).
+On the other hand, dst_arg_index=-1 is useful when inputs have different node types and allowed switching positions like
```

Comment:
*matching

Would also be good to document the validity/usage of -1 a bit more clearly, perhaps as a comment in the definition of MatchEdgeInfo.dst_arg_index as it's only indirectly documented here in the remarks.